api/container.gql

```diff
@@ -22,17 +22,28 @@ Examples:
 scalar ContainerAddress
 
 """
-"An OCI-compatible container, also known as a "docker container"
+An OCI-compatible container, also known as a "docker container"
 """
 type Container {
     "A unique identifier for this container"
     id: ContainerID!
 
-    "Initialize this container from the base image published at the given address"
+    "Initialize this container to the base image published at the given address"
     from(address: ContainerAddress!): Container!
 
+    """
+    Initialize this container to the result of the specified build.
+
+    The build can be customized to use a Dockerfile at a non-standard location,
+    or (in the future) to support files other than Dockerfile.
+    """
+    build(source: DirectoryID!, file: String="./Dockerfile")
```

Comment:
IMHO this should be part of directory as it always takes a directory as first argument.

It's the same principle we've applied to all other fields. In principle, these should be a mutation taking a container or directory as first argument. But we've decided instead to have them as fields of the type they would usually take as argument, to allow chaining.

In this case it's neither a mutation nor it allows chaining.

e.g.

```graphql
git {
  build { # <-- can't be done
  }
}
```

```graphql
{ 
  container {
    exec(...) {
      directory(path: "/out) {
        build { # can't be done
        }
      }
    }
  }
}
```

Reply:
I think this is a one-way door that we are not yet ready to cross, because it enshrines the concept of extensions being allowed / expected to add higher-level fields (like `dockerBuild`) to lower-level types (like `Directory`), something we are not yet ready to decide on IMO.

Yes our docker build implementation is built-in, but it sets a clear precedent for adding more build methods down the road:

* Why `Directory { build }` and not `Directory { dockerBuild }`? After all there are other ways to build from a directory.
* If `Directory { dockerBuild }`, why not `Directory { yarnBuild }`?
* etc.

This may in fact be the pattern we choose for extensions. My point is that `Directory { build }` would force us into a decision too early.

The downside is that we lose an opportunity to chain in certain queries, but that constraints already exists anyway (and we can always cross the one-way door later).

Reply:
One problem is that implementing docker build with a container will not be very straightforward as you can't inherently provide mounts alongside the context you send to the dockerfile frontend.

You could simulate normal read-write directory mounts, but you'd probably just have to ignore cache+tmpfs mounts entirely (or doing something truly insane like rewrite the Dockerfile to always include those with the context fs, which just isn't really viable imo).

A possible intermediate solution is to not extend Container or Directory with this and instead just add a top-level `dockerbuild(dir: DirectoryID!)`. I don't like that either to be honest, but it at least would skirt around having to make a decision here.